### ISSUE TITLE
storage: prefix bloom filter to avoid unneccessary scan for non-existent join keys

### ISSUE THREAD

**Issue by hzxa21** [singularity-data/risingwave#2947]
We observe intensive reads to storage on new keys in join in a recent benchmark conducted by @skyzh.

We are exploring options to improve it and one fesible solution is to introduce a bloom filter for a the join key so that we can avoid unneccessary I/O for new keys. However, the join key can be variable-length. In general, we don't want the concept of table schema to be intrusive to our KV storage. 

Per our slack discussion, one way to solve the issue is to store the prefix length in value meta and build a bloom filter of variable-length prefix in SST.

https://singularity-data.slack.com/archives/C03GTKX3C8G/p1653468068380959
> I didn’t get why the schema is needed? Can we just tell the storage the prefix length should be used in bloom filter when insert? (or even put it in value meta)
>> Sounds possible. Besides, if prefix length exists in value meta, the vnode is not necessary because it can be computed from distribution key.
>>> Good idea! We just need to understand prefix length, not the schema in this case

Open for discussions!

**Comment by fuyufjh** [comment]
I would +1 for letting the storage be aware of schema i.e. stores the schema for every table ID 🤪 Here are the pros:

1. The prefix length is another kind of value meta, so every key-value entry would pay this cost. While for a schema-aware storage, this is not needed.
2. The table schema scan also is used to calculate `vnode`, so `vnode` (2 bytes) can also be omitted from value meta.
3. (Potentially) Schema is helpful if you want to build any index during compaction, as we have planned in very early discussion.
4. Anyway, the `CellBasedTable` interface IS aware of schema now, so I think it looks not so wired to do such assumption.

BTW, as you may know, B-Tree-based databases use comparison (`row1 < row2`) to maintain the structure of search tree, so the B+Tree must understand the schema, too.

**Comment by hzxa21** [comment]
> Is it possible to do it on the executor level, i.e. HashJoinExecutor, I guess not customizing things into the storage is preferred and can be tried first, or this prefix bloom filter can also be enabled for other scenarios.

+1 for doing optimizations on executor level first. In terms of whether prefix bloom filter is beneficial in other cases, anyone who does prefix scan on storage can be the candidate. For exmaple, I think it can also help scaning TopN states to check existence for new group keys.

**Comment by jon-chuang** [comment]
So I have some questions:
1. Is the bloom filter an additional prefix bloom filter built in addition to point bloom filter? Or is it an alternative? For instance, certain tables may only want to perform key scan, but not point lookups. Could we configure the bloom filter of a given table to always be prefix filter?
2. I feel that prefix filter is something that is opt-in, where the default is just point filter. So it doesn't seem to hurt to provide the table schema when opting-in to a prefix bloom filter. So rather than store table schema, we could just do so for those tables that require prefix filter.

**Comment by fuyufjh** [comment]
> For my comments: I feel that prefix filter is something that is opt-in, where the default is just point filter. So it doesn't seem to hurt to provide the table schema when opting-in to a prefix bloom filter. So rather than store table schema in all tables, we could just do so for those tables that require prefix filter as part of the config to enable it.

Actually, for our use cases, "point filter" (bloom-filter) is almost useless because of both epoch and operator's behavior.

1. **Epoch**. In Hummock, an epoch (aka. timestamp in other systems) is always stored as a suffix to user key. However, you can never know the epoch when reading, but only scan all the keys and returns the latested one.
2. **Operator's behavior**. In streaming, stateful operators like HashJoin, HashAgg and TopN mostly uses range scan instead of point gets. For example, if cache-miss happended, `HashJoin` always has to scan all the keys for a single join key. @wcy-fdu once had a benchmark on this and we also saw this fact.

In summary, I think a classical bloom-filter is useless and I think it can be just replaced by our customized bloom-filter of the actual user key.

**Comment by hzxa21** [comment]
> > For my comments: I feel that prefix filter is something that is opt-in, where the default is just point filter. So it doesn't seem to hurt to provide the table schema when opting-in to a prefix bloom filter. So rather than store table schema in all tables, we could just do so for those tables that require prefix filter as part of the config to enable it.
> 
> Actually, for our use cases, "point filter" (bloom-filter) is almost useless because of both epoch and operator's behavior.
> 
> 1. **Epoch**. In Hummock, an epoch (aka. timestamp in other systems) is always stored as a suffix to user key. However, you can never know the epoch when reading, but only scan all the keys and returns the latested one.
> 2. **Operator's behavior**. In streaming, stateful operators like HashJoin, HashAgg and TopN mostly uses range scan instead of point gets. For example, if cache-miss happended, `HashJoin` always has to scan all the keys for a single join key. @wcy-fdu once had a benchmark on this and we also saw this fact.
> 
> In summary, I think a classical bloom-filter is useless and it can be just replaced by our customized bloom-filter of a "filtering key" defined for each table ID.

Currently the bloom-filter is generated based on the user key, not full key so 1) is not an issue. It is true that the current bloom-filter can only accelerate point-get on the full user key so it only benefits the `get`. `scan` cannot use the current bloom-filter so 2) is valid.

**Comment by hzxa21** [comment]
@Little-Wallace points out that we can allow upper layer (e.g. relational layer) to provide a funtion (`extract_bloom_filter_key`) to storage so that storage can extract the bloom filter key without understanding the schema. For example, for join on a varchar join key, JoinStateTable can register `extract_bloom_filter_key` with the implementation as looking up the join table schema and parsing the real join key from the storage key.

This no only helps building prefix bloom filter but also can extend to support building bloom filter for arbitrary transformation of a storage key (of course with the extra cost to transform a key).

**Comment by skyzh** [comment]
That's quite costly, as we will need to decode the key by schema (which will further be provided by meta).

> The prefix length is another kind of value meta

Is my preferred solution.

**Comment by fuyufjh** [comment]
> we can allow upper layer (e.g. relational layer) to provide a funtion (extract_bloom_filter_key) to storage so that storage can extract the bloom filter key without understanding the schema.

But how to pass the function to compactor? 🤔

I think it's identical to my proposal ("letting the storage be aware of schema") in implementation.

**Comment by fuyufjh** [comment]
> That's quite costly, as we will need to decode the key by schema (which will further be provided by meta).
> 
> > The prefix length is another kind of value meta
> 
> Is my preferred solution.

It's too expensive for our cell-based storage. 😇 I would rather revert back to the KV-per-row design.

**Comment by hzxa21** [comment]
Another idea (from @skyzh): any executor that needs to perform prefix scan writes a entry with the prefix as the key. In this case, we can use the regular bloom filter to filter out non-existent prefix.

An extension to this approach is only consider the prefix-key in the bloom filter, not other keys to make our bloom filter more effiecient. This makes sense because for some executors (e.g. Join, TopN), they will never do point-get from storage so maintaining bloom filter for each key is useless.

**Comment by fuyufjh** [comment]
> Another idea (from @skyzh): any executor that needs to perform prefix scan writes a entry with the prefix as the key. In this case, we can use the regular bloom filter to filter out non-existent prefix.
> 
> An extension to this approach is only consider the prefix-key in the bloom filter, not other keys to make our bloom filter more effiecient. This makes sense because for some executors (e.g. Join, TopN), they will never do point-get from storage so maintaining bloom filter for each key is useless.

Didn't get it actually. How to let the compactor know the what's the prefix? 😇

**Comment by hzxa21** [comment]
> > Another idea (from @skyzh): any executor that needs to perform prefix scan writes a entry with the prefix as the key. In this case, we can use the regular bloom filter to filter out non-existent prefix.
> > An extension to this approach is only consider the prefix-key in the bloom filter, not other keys to make our bloom filter more effiecient. This makes sense because for some executors (e.g. Join, TopN), they will never do point-get from storage so maintaining bloom filter for each key is useless.
> 
> Didn't get it actually. How to let the compactor know the what's the prefix? 😇

Right now join state is stored with key `table_id | join_key | pk | col_id`. Join executor can write one extra kv entry for each join_key with `table_id | join_key` as the storage key and we can build a normal bloom filter as usual for this key.

**Comment by lmatz** [comment]
a tombstone-like entry

**Comment by fuyufjh** [comment]
> Right now join state is stored with key `table_id | join_key | pk | col_id`. Join executor can write one extra kv entry for each join_key with `table_id | join_key` as the storage key and we can build a normal bloom filter as usual for this key.

It's specified to operator, right? If so, exposing schema is more clean to me.

**Comment by hzxa21** [comment]
> I am interested in this issue, may I have a try on this？I will create a tracking issue describing the steps to implement it after discuss with wallace

Sure. You can self-assign.
